src/nodes/Table/markdown.ts

```diff
@@ -0,0 +1,202 @@
+/**
+ * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+import { MarkdownSerializerState } from '@tiptap/pm/markdown'
+import { Node } from '@tiptap/pm/model'
+import { createMarkdownSerializer } from '../../extensions/Markdown.js'
+
+type Cell = {
+	md: string
+	lines: string[]
+	nodeTypes: Set<string>
+	align: string
+}
+
+type Row = {
+	node: Node
+	header: boolean
+	cells: Cell[]
+	length: number
+}
+
+// Allowed node types for center/right alignment. Cells with any other node type have
+// enforced left alignment to not break markdown parsing.
+const alignNodeTypes = new Set(['text', 'paragraph', 'image'])
+
+// Block node types that may span only a single line.
+const singleLineBlockNodeTypes = new Set([
+	'blockquote',
+	'bulletList',
+	'heading',
+	'orderedList',
+	'taskList',
+])
+
+/**
+ * Serialize a table row to markdown line by line
+ *
+ * @param state - Markdown serializer state
+ * @param row - Table row to serialize
+ * @param columnWidths - Array of column widths
+ */
+function rowToMarkdown(
+	state: MarkdownSerializerState,
+	row: Row,
+	columnWidths: number[],
+) {
+	const normalizedCells = row.cells.map((cell, cellIdx) => {
+		// Append newline to single-line cells with a block node
+		if (
+			cell.lines.length === 1
+			&& [...cell.nodeTypes].some((type) => singleLineBlockNodeTypes.has(type))
+		) {
+			cell.lines.push('')
+			row.length = Math.max(row.length, 2)
+		}
+
+		// Normalize cells to have the same number of lines
+		while (cell.lines.length < row.length) cell.lines.push('')
+		// Normalize lines in cell to have the same length
+		cell.lines.forEach((line, lineIdx) => {
+			if ([...cell.nodeTypes].some((type) => !alignNodeTypes.has(type))) {
+				// Enforced left alignment.
+				cell.lines[lineIdx] = line.padEnd(columnWidths[cellIdx])
+				return
+			}
+
+			// Pad according to alignment
+			if (cell.align === 'center') {
+				const spaces = Math.max(columnWidths[cellIdx] - line.length, 0)
+				const spacesStart = line.length + Math.floor(spaces / 2)
+				const spacesEnd = line.length + Math.ceil(spaces / 2)
+				cell.lines[lineIdx] = line.padStart(spacesStart).padEnd(spacesEnd)
+			} else if (cell.align === 'right') {
+				cell.lines[lineIdx] = line.padStart(columnWidths[cellIdx])
+			} else {
+				cell.lines[lineIdx] = line.padEnd(columnWidths[cellIdx])
+			}
+		})
+		return cell
+	})
+
+	// Write row line by line
```

Comment:
I'd make this a separate function. Both in terms of what they do and which arguments they need they can be separated nicely. The part above does not touch `state` at all the part below has nothing to do with `columnWidths`.

Reply:
I outsourced the `normalizeCells()` part into an own function now, hope that's what you had in mind: 12cb94a2246fd18d61fc2b6fe5c09c16f8532811